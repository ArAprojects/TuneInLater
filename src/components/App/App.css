main {
  height: 100vh;
  background: url('./book4.jpg');
  background: linear-gradient(rgba(0, 0, 0, 0.4), rgba(0, 0, 0, 0)), url('./book4.jpg');
  background-attachment: fixed;
  background-size: cover;
  background-position: bottom;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  overflow: auto;
}

* {
  font-family: Optima, sans-serif;
}
