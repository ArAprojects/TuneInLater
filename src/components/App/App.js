import React, { Component } from 'react';
import Nav from '../Nav/Nav';
import BookContainer from '../BookContainer/BookContainer';
import Login from '../Login/Login';
import { Route } from 'react-router-dom'
import './App.css';
import api from '../API/api.js'

class App extends Component {
  constructor() {
    super()
    this.state = {
      audiobooks: [],
      currentUser: '',
      favorites: []
    }
  }

  newSearch = async (text) => {
    const audio = await api.getAudio(text)
    this.setState({audiobooks: audio})
  }

  makeNewUser = async (userInfo) => {
    await api.createNewUser(userInfo)
  }

  logInUser = async (userInfo) => {
    const current = await api.logIn(userInfo)
    this.setState({ currentUser: current})
  }

  postFavorite = async (favoriteInfo, currentUserID) => {
    await api.newFavorite(favoriteInfo, currentUserID)
  }


// bookfavorites requires: book_id (Integer), author_name (String),
// book_name VARCHAR (String), artwork_url (String), release_date (String),
// description (String), primary_genre_name (String)

  structureObject = (favorite) => {
    return {
      book_id: favorite.book_id,
      author_name: favorite.author_name,
      book_name: favorite.book_name,
      artwork_url: favorite.artwork_url,
      release_date: favorite.release_date,
      description: favorite.description,
      primary_genre_name: favorite.primary_genre_name
    }
  }


  toggleFavorite = (favorite) => {
    console.log(favorite)
    if (this.state.favorites.find(book => book.book_name === favorite.book_name)) {
      let index = this.state.favorites.indexOf(favorite);
      this.state.favorites.splice(index, 1);
      this.setState({...this.state.favorites})
      favorite["favorite"] = false;
    } else {
      favorite.favorite = true;
      this.setState({ favorites: [...this.state.favorites, favorite] });
      const favoriteBook = this.structureObject(favorite)
      this.postFavorite(favoriteBook, this.state.currentUser.id)
    }

}

  render () {
    console.log('app state', this.state)
    const { audiobooks } = this.state
    return (
      <div>
        <Route exact path='/' render={() => <Login loginUser={this.logInUser} createNewUser={this.makeNewUser} /> } />
        <Route path='/home' render={() =>
          <main>
            <Nav newSearch={this.newSearch} currentUser={this.state.currentUser} />
            <BookContainer audiobooks={audiobooks} toggleFavorite={this.toggleFavorite} />
          </main>
        } />
      </div>
    )
  }
}

export default App;
