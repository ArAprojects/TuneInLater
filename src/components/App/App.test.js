import React from 'react';
import { shallow } from 'enzyme';
import { App, mapStateToProps } from './App';

describe('App', () => {
  let wrapper;
  const mockSelectUser = jest.fn();
  const mockGetAudio = jest.fn();
  const mockToggleFavorite = jest.fn();
  const mockGetUserFavofites = jest.fn();
  const mockCurrentUser = { 
    id: 3, 
    name: "bill", 
    email: "bill@gmail.com" 
  }
  const mockAudioBooks = [
    {
      artwork_url: 'http//:image.png',
      author_name: 'Dr. Seuss',
      book_id: 1,
      book_name: 'The Cat in the Hat',
      description: 'Story Description',
      favorite: true,
      primary_genre_name: 'children',
      release_date: '04/23/1990'
    },
    {
      artwork_url: 'http//:image.png',
      author_name: 'Dr. Seuss',
      book_id: 2,
      book_name: 'Green Eggs & Ham',
      description: 'Story Description',
      favorite: false,
      primary_genre_name: 'children',
      release_date: '04/23/1990'
    }
  ]
  const mockFavoriteBooks = [
    {
      artwork_url: 'http//:image.png',
      author_name: 'Dr. Seuss',
      book_id: 1,
      book_name: 'The Cat in the Hat',
      description: 'Story Description',
      favorite: false,
      primary_genre_name: 'children',
      release_date: '04/23/1990'
    }
  ]

  beforeEach(() => {
    wrapper = shallow(<App 
      selectCurrentUserReducer={mockCurrentUser}
      getAudiobooksReducer={mockAudioBooks}
      favoritesReducer={mockFavoriteBooks}
      selectCurrentUser={mockSelectUser}
      getAudiobooks={mockGetAudio}
      toggleFavoriteBook={mockToggleFavorite}
      getUserFavorites={mockGetUserFavofites}
    />)
  })

  it('should take a snapshot when passed the correct data', () => {
    expect(wrapper).toMatchSnapshot();
  })

  

  describe('mapStateToProps', () => {
    it('should return the current user, an array of books, and an array of favorite books', () => {
      const mockState = {
        selectCurrentUserReducer: { id: 1, name: 'Jacob' },
        getAudiobooksReducer: ['book1', 'book2'],
        toggleFavoriteReducer: ['book2'],
        toggleFavoriteBtnReducer: 'favorite'
      };
      const expected = {
        selectCurrentUserReducer: { id: 1, name: 'Jacob' },
        getAudiobooksReducer: ['book1', 'book2'],
        favoritesReducer: ['book2']     
      };

      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
  });

  // describe('mapDispatchToProps', () => {
  //   it('calls dispatch with an addTodo action when handleSubmit is called', () => {
  //     // Setup
  //     const mockDispatch = jest.fn();
  //     const actionToDispatch = addTodo('Learn Redux!', 1);

  //     // Execution
  //     const mappedProps = mapDispatchToProps(mockDispatch);
  //     mappedProps.handleSubmit('Learn Redux!', 1);

  //     // Expectaion
  //     expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  //   });
  // });
})