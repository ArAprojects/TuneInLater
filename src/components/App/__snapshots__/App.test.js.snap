// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should take a snapshot when passed the correct data 1`] = `
<div
  className="the-app"
>
  <Route
    exact={true}
    path="/login"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/favorites"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/details/:id"
    render={[Function]}
  />
</div>
`;
